pennylane/devices/qubit/sampling.py

```diff
@@ -213,6 +217,97 @@ def measure_with_samples(
     return sorted_res
 
 
+def _measure_with_samples_nan(
+    mps: List[SampleMeasurement], state: np.ndarray, shots: Shots, is_state_batched: bool = False
+) -> TensorLike:
+    batch_size = qml.math.shape(state)[0] if is_state_batched else None
+    interface = qml.math.get_interface(state)
+
+    if shots.has_partitioned_shots:
+        # res = []
+        # for s in shots:
+        #     res.extend(_get_single_shot_nan_res(mps, batch_size, interface, s))
+        # res = tuple(res)
+        res = tuple(_get_single_shot_nan_res(mps, batch_size, interface, s) for s in shots)
+    else:
+        res = _get_single_shot_nan_res(mps, batch_size, interface, shots.total_shots)
+
+    return res
+
+
+def _get_single_shot_nan_res(mps, batch_size, interface, shots):
```

Comment:
I need to understand more here, but I worry that there is too much handling of the `NaN` edge case and I wonder how much we can just rely on the interface to propagate the `NaN`s.